### ISSUE TITLE
It's hard for py devs to see contract methods of DataNFT, Datatoken, etc

### ISSUE THREAD

**Issue by trentmc** [oceanprotocol/ocean.py#1165]
### Motivation

Recently, ocean.py got brownie support. This meant we could (and did) deprecate all the ocean.py methods that simply passed-through to smart contracts. 

For example, Datatoken.balanceOf(). And probably 200 other methods across modules in `ocean_lib/models/*.py`.

This _greatly_ reduced the amount of code to maintain & debug. It also made a smoother UX.

However, we _did_ lose one thing. Once developers go beyond the ocean.py READMEs, they don't have a great reference for what methods are available. They must resort to (a) looking at unit tests (b) looking at solidity code.

Can we...
- help developers to understand the smart contract methods that DataNFT, Datatoken, etc expose
- while maintaining the benefits of brownie

### Suggested Approach

For each module / class in `ocean_lib/models/*.py`, add a triple-quote string `"""foo bar"""` that spans many lines. Inside it, put the method definitions for each smart contract method, written in a py-like fashion (to maximize approachability to py devs).

E.g. for datatoken.py's Datatoken class:
```python

class Datatoken:
   # have the usual non-smart-contract methods first. Then...
 
"""
  def balanceOf(address: str) -> int:
    Returns the balance of this address.
    :param: address of interest
    :return: balance, in wei 

  def approve(...
"""
   
```

**Comment by trentmc** [comment]
Note: the fix for #1095 included a new main flow, and its appendix nicely references what interfaces developers can use for data NFTs and datatokens. That will help soften the concern of this issue.

However, we should still do as this issue prescribes: a docstring in each py module listing the smartphone contract interfaces. That will help further.

**Comment by calina-c** [comment]
I recommend `@method` annotations instead. https://manual.phpdoc.org/HTMLframesConverter/default/phpDocumentor/tutorial_tags.method.pkg.html

**Comment by trentmc** [comment]
> I recommend something like `@method` annotations instead. https://manual.phpdoc.org/HTMLframesConverter/default/phpDocumentor/tutorial_tags.method.pkg.html
> 
> There has to be something similar in Python!

I [commented in the PR](https://github.com/oceanprotocol/ocean.py/pull/1270#issuecomment-1377133706).

**Comment by calina-c** [comment]
@alexcos20 @trentmc I have completed the first step, an inventory of the methods needing documentation. I have some questions, so I'll include the report below. Please help with any information/guidance you may see fit.

### General issues
* `info` method  is added by default, but is now empty for our contracts, should we document/keep it? This applies to all contracts
* For other default methods i.e. `events`, `topics` should we add docstrings? These are generic methods on all contracts. Should we add docstrings in each model file to present them or is there a better way?
* `router` is a similar case, but not present on *ALL* contracts
* when a contract function is wrapped in a helper, I did *not* include it with the docstrings. i.e. for FixedRate `setRate` is abstracted by us in OneExchange as `set_rate`, in Datatoken the `startOrder` is wrapped by  `start_order` which handles tuple arguments and some defaults. I would rather not include the raw methods as docstrings if we do wrap them. Having both options would only lead to duplication, confusion and mistakes. However, I need confirmation of this train of thought.

### Data NFT Factory issues
* as far as I know, we should not allow users to add, enable, nor disable token templates via ocean.py. I propose not documenting these functions and, preferably, even pruning them from the list of transferred functions.
* `createNftWithErc20WithPool` -> should we even keep this in the contracts at all? As far as I know, pools are not used anymore. Should we document it? Should we prune it from the model?

### Datatoken issues
* functions related to pools i.e. `deployPool` and `getPools` -> see similar issue above, same questions

### DatatokenEnterprise issues
* same issue with pool functions (`getPools`)
* do we have to copy/paste the docstrings from Datatoken, or can I just document the functions that are different? The documentation is the same in both cases for ~20 functions

### ERC721TokenFactoryBase issues
* this is not a model per se, but a base class. Should I also add docstrings here? I think adding them on the data NFT factory is enough.

### Factory Router issues
* should we allow any write calls to the router? This class seems to have a lot of outdated code like that for pools and SS contracts. I can provide the entire list of functions to continue the discussion if needed.
* should I add docstrings for constants? e.g. `MAX_FEE`, `MIN_WEIGHT` etc?

### FRE issues
* should I add docstrings for constants? e.g `MAX_FEE`, `MIN_FEE` etc? If so, should they be added as wrappers in OneExchange, or docstrings on the FRE model itself?
* should we implement a wrapper for `createWithDecimals` ? I don’t know a lot about this function and I was wondering if our FRE code supports this or it should be just a docstring with raw usage options
* we have a wrapper for `getExchange` using the OneExchange class. Should we implement a similar wrapper for `getExchanges`?

**Comment by trentmc** [comment]
Answers inline.

> ### General issues
> * `info` method  is added by default, but is now empty for our contracts, should we document/keep it? This applies to all contracts

Don't document it.

> * For other default methods i.e. `events`, `topics` should we add docstrings? These are generic methods on all contracts. Should we add docstrings in each model file to present them or is there a better way?

I'd say: skip here. Too much detail for this format. We can consider adding them as part of good-docs issue [#1184](https://github.com/oceanprotocol/ocean.py/issues/1184).

> * `router` is a similar case, but not present on _ALL_ contracts

Like above.

> * when a contract function is wrapped in a helper, I did _not_ include it with the docstrings. i.e. for FixedRate `setRate` is abstracted by us in OneExchange as `set_rate`, in Datatoken the `startOrder` is wrapped by  `start_order` which handles tuple arguments and some defaults. I would rather not include the raw methods as docstrings if we do wrap them. Having both options would only lead to duplication, confusion and mistakes. However, I need confirmation of this train of thought.

I think it's useful for people to be aware of these, just with lower priority. I'd put them at the bottom, and mark as "redundant with __".

> ### Data NFT Factory issues
> * as far as I know, we should not allow users to add, enable, nor disable token templates via ocean.py. I propose not documenting these functions and, preferably, even pruning them from the list of transferred functions.

You can skip documenting these. Don't bother pruning from transferred functions.

> * `createNftWithErc20WithPool` -> should we even keep this in the contracts at all? As far as I know, pools are not used anymore. Should we document it? Should we prune it from the model?

Don't bother documenting. Don't bother pruning.

> ### Datatoken issues
> * functions related to pools i.e. `deployPool` and `getPools` -> see similar issue above, same questions

Don't bother documenting. Don't bother pruning.

> ### DatatokenEnterprise issues
> * same issue with pool functions (`getPools`)

Don't bother documenting. Don't bother pruning.

> * do we have to copy/paste the docstrings from Datatoken, or can I just document the functions that are different? The documentation is the same in both cases for ~20 functions

Follow DRY.

> ### ERC721TokenFactoryBase issues
> * this is not a model per se, but a base class. Should I also add docstrings here? I think adding them on the data NFT factory is enough.

Yes, do it just on data nft factory.

> ### Factory Router issues
> * should we allow any write calls to the router? This class seems to have a lot of outdated code like that for pools and SS contracts. I can provide the entire list of functions to continue the discussion if needed.

Skip it.

> * should I add docstrings for constants? e.g. `MAX_FEE`, `MIN_WEIGHT` etc?

I'd say: skip here. Like above, add later as part of good-docs.

> ### FRE issues
> * should I add docstrings for constants? e.g `MAX_FEE`, `MIN_FEE` etc? If so, should they be added as wrappers in OneExchange, or docstrings on the FRE model itself?

I'd say: skip here. Like above, add later as part of good-docs.

> * should we implement a wrapper for `createWithDecimals` ? I don’t know a lot about this function and I was wondering if our FRE code supports this or it should be just a docstring with raw usage options

I don't know a lot about `createWithDecimals` either. Maybe it should be deprecated overall? Perhaps take a few min to explore further, and then if appropriate, create a github issue.

> * we have a wrapper for `getExchange` using the OneExchange class. Should we implement a similar wrapper for `getExchanges`?

Perhaps take a few min to explore further, and then if appropriate, create a github issue.

**Comment by calina-c** [comment]
Thank you, Trent! I've addressed almost everything locally. I'm commenting with a task list of what is pending.:

- [ ] document redundant functions (those wrapped in helpers) at the bottom of the docstrings
- [ ] follow-up on createWithDecimals and getExchanges()

btw I sincerely believe we shouldn't allow users to mess with the router at all :) maybe we could skip the read functions as well. In fact, in almost all cases we are only using the router address through ocean.py. Anything non-address is only used as part of test_factory_router.py. We can keep the tests, but I don't think functions regarding the router should be touched through ocean.py. CC: @trentmc @alexcos20

**Comment by trentmc** [comment]
> btw I sincerely believe we shouldn't allow users to mess with the router at all :) maybe we could skip the read functions as well.

I'm ok with skipping router functionality here.

**Comment by calina-c** [comment]
Closed by #1270
